### PR TITLE
MODFISTO-62 Payment restrictions

### PR DESCRIPTION
**Pull request by grabsefx**
# Purpose
https://issues.folio.org/browse/MODFISTO-62

## Approach
* implemented payment restrictions
* unit tests updated

#### TODOS and Open Questions
* It looks like there is a problem with [calculation formula](https://wiki.folio.org/display/FOLIJET/Finance+Data+Model#FinanceDataModel-TransactionRestrictions). Wait until it will be corrected

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java [hunk 1]
```diff
@@ -376,7 +376,8 @@ protected String getValues(List<JsonObject> entities) {
 
   private Future<Void> checkEncumbranceRestrictions(Transaction transaction, Budget budget) {
     Promise<Void> promise = Promise.promise();
-    if (transaction.getTransactionType() == Transaction.TransactionType.ENCUMBRANCE) {
+    if (transaction.getTransactionType() == Transaction.TransactionType.ENCUMBRANCE
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java]: I think Credits can be applied only to `Active` budgets as well. @craigmcnally  is that correct?

File: src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java [hunk 2]
```diff
@@ -384,7 +384,8 @@ protected String getValues(List<JsonObject> entities) {
 
   private Future<Void> checkEncumbranceRestrictions(Transaction transaction, Budget budget) {
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java]: Please give this method a more general name, as it checks not only encumbrances, but also payments.

File: src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java [hunk 3]
```diff
@@ -440,6 +458,30 @@ private Double getBudgetRemainingAmountForEncumbrance(Budget budget, String curr
     return result.getNumber().doubleValue();
   }
 
+  /**
+   * Calculates remaining amount for payment
+   * [remaining amount] = (allocated * allowableExpenditure) - (allocated - (unavailable + available)) - (awaitingPayment + expended)
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java]: @craigmcnally Perhaps we should consider the encumbered value in this formula. For example, it is possible that we are trying to make a payment from a budget whose money is encumbered for orders not associated with this payment.

File: src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java [hunk 4]
```diff
@@ -416,6 +424,16 @@ protected String getValues(List<JsonObject> entities) {
...
+  }
+
+  private boolean isPaymentRestricted(Transaction transaction, Budget budget, Ledger ledger) {
+    return transaction.getTransactionType() == Transaction.TransactionType.PAYMENT && ledger.getRestrictExpenditures()
+      && budget.getAllowableExpenditure() != null;
+  }
+
   /**
    * Calculates remaining amount for encumbrance
    * [remaining amount] = (allocated * allowableEncumbered) - (allocated - (unavailable + available)) - (encumbered + awaitingPayment)
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java]: Could you update calculation of remaining amount for encumbrance according to discussion on [WIKI](https://wiki.folio.org/display/FOLIJET/Finance+Data+Model#FinanceDataModel-Encumbrance)

File: src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java (hunk 2 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java]: Sure. Naming updated

File: src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java (hunk 4 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java]: added expenditures into calculations formula

File: src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java (hunk 1 above)
Comment (craigmcnally) [reply src/main/java/org/folio/rest/transaction/AllOrNothingHandler.java]: That's correct.  Apologies for the delay

**Pull request merged**